protocols/rendezvous/src/codec.rs

```diff
@@ -385,118 +371,110 @@ impl TryFrom<wire::Message> for Message {
     fn try_from(message: wire::Message) -> Result<Self, Self::Error> {
         use wire::message::*;
 
-        let message = match message {
-            wire::Message {
-                r#type: Some(0),
-                register:
-                    Some(Register {
-                        ns,
-                        ttl,
-                        signed_peer_record: Some(signed_peer_record),
-                    }),
-                ..
-            } => Message::Register(NewRegistration {
-                namespace: ns
-                    .map(Namespace::new)
-                    .transpose()?
-                    .ok_or(ConversionError::MissingNamespace)?,
-                ttl,
-                record: PeerRecord::from_signed_envelope(SignedEnvelope::from_protobuf_encoding(
-                    &signed_peer_record,
-                )?)?,
-            }),
-            wire::Message {
-                r#type: Some(1),
-                register_response:
-                    Some(RegisterResponse {
-                        status: Some(0),
-                        ttl,
-                        ..
-                    }),
-                ..
-            } => Message::RegisterResponse(Ok(ttl.ok_or(ConversionError::MissingTtl)?)),
-            wire::Message {
-                r#type: Some(3),
-                discover: Some(Discover { ns, limit, cookie }),
-                ..
-            } => Message::Discover {
-                namespace: ns.map(Namespace::new).transpose()?,
-                cookie: cookie.map(Cookie::from_wire_encoding).transpose()?,
-                limit,
-            },
-            wire::Message {
-                r#type: Some(4),
-                discover_response:
-                    Some(DiscoverResponse {
-                        registrations,
-                        status: Some(0),
-                        cookie: Some(cookie),
-                        ..
-                    }),
-                ..
-            } => {
-                let registrations = registrations
-                    .into_iter()
-                    .map(|reggo| {
-                        Ok(Registration {
-                            namespace: reggo
-                                .ns
+        let msg_ty = match message.type_ {
+            Some(ty) => ty.enum_value().or(Err(ConversionError::InconsistentWireMessage))?,
+            None => return Err(ConversionError::InconsistentWireMessage),
+        };
+
+        let message = match msg_ty {
+            MessageType::REGISTER => {
+                match message.register.into_option() {
```

Comment:
I think this may be cleaner too if we'd use `ok_or(ConversionError::InconsistentWireMessage)?` here and unpack the fields one by one instead of pattern matching on it.